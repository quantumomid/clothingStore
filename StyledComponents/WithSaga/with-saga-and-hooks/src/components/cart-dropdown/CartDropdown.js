import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { useHistory } from "react-router-dom"
import { selectCartItems } from "../../redux/cart/cartSelectors";
import CartItem from "../cart-item/CartItem";
import { toggleCartHidden } from "../../redux/cart/cartActions";
import { CartDropdownButton, CartDropdownContainer, CartItemsContainer, EmptyMessageContainer } from "./cartDropdownStyles";

const CartDropdown = () => {
    const dispatch = useDispatch();
    const cartItems = useSelector(selectCartItems);

    const history = useHistory();
    return (
        <CartDropdownContainer>
            <CartItemsContainer>
            {   cartItems.length ? (
                cartItems.map(cartItem => <CartItem key={cartItem.id} item={cartItem} />))
                : <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
            }
            </ CartItemsContainer>
            <CartDropdownButton onClick={() => {
                history.push('/checkout')
                dispatch(toggleCartHidden())   
            }}>GO TO CHECKOUT</CartDropdownButton>
        </ CartDropdownContainer>
    )
};

// Note: the order of wrapping is important i.e. to pass the "connected" component
// into withRouter so that the correct end state of CartDropdown has access to the 
// various router properties we desire
export default CartDropdown;
