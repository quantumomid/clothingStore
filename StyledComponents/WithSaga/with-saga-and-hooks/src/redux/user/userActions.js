import { EMAIL_SIGN_IN_START, SIGN_IN_SUCCESS, SIGN_IN_FAILURE, GOOGLE_SIGN_IN_START, CHECK_USER_SESSION, SIGN_OUT_START, SIGN_OUT_SUCCESS, SIGN_OUT_FAILURE, REGISTER_START, REGISTER_SUCCESS, REGISTER_FAILURE } from "./userActionTypes"

export const googleSignInStart = () => ({
    type: GOOGLE_SIGN_IN_START
})

export const signInSuccess = (user) => ({
    type: SIGN_IN_SUCCESS,
    payload: user
})

export const signInFailure = (error) => ({
    type: SIGN_IN_FAILURE,
    payload: error
})

export const emailSignInStart = (emailAndPassword) => ({
    type: EMAIL_SIGN_IN_START,
    payload: emailAndPassword
})

export const checkUserSession = () => ({
    type: CHECK_USER_SESSION
})

export const signOutStart = () => ({
    type: SIGN_OUT_START
})

export const signOutSuccess = () => ({
    type: SIGN_OUT_SUCCESS
})

export const signOutFailure = (error) => ({
    type: SIGN_OUT_FAILURE,
    payload: error
})

export const registerStart = (userRegisterDetails) => ({
    type: REGISTER_START,
    payload: userRegisterDetails
})

export const registerSuccess = ({ user, additionalData }) => ({
    type: REGISTER_SUCCESS,
    payload: { user, additionalData }
})

export const registerFailure = (error) => ({
    type: REGISTER_FAILURE,
    payload: error
})