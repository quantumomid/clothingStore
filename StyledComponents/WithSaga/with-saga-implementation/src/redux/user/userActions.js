import { EMAIL_SIGN_IN_START, SIGN_IN_SUCCESS, SIGN_IN_FAILURE, GOOGLE_SIGN_IN_START, SET_CURRENT_USER, CHECK_USER_SESSION } from "./userActionTypes"

export const googleSignInStart = () => ({
    type: GOOGLE_SIGN_IN_START
})

export const signInSuccess = (user) => ({
    type: SIGN_IN_SUCCESS,
    payload: user
})

export const signInFailure = (error) => ({
    type: SIGN_IN_FAILURE,
    payload: error
})

export const emailSignInStart = (emailAndPassword) => ({
    type: EMAIL_SIGN_IN_START,
    payload: emailAndPassword
})

export const checkUserSession = () => ({
    type: CHECK_USER_SESSION
})
