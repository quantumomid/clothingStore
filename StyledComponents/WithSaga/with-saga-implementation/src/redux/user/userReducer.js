import { EMAIL_SIGN_IN_FAILURE, EMAIL_SIGN_IN_SUCCESS, GOOGLE_SIGN_IN_FAILURE, GOOGLE_SIGN_IN_SUCCESS, SET_CURRENT_USER } from "./userActionTypes"

const initialState = {
    currentUser: null,
    error: null
}

const userReducer = (state = initialState, action) => {

    switch(action.type){
        case GOOGLE_SIGN_IN_SUCCESS:
        case EMAIL_SIGN_IN_SUCCESS:
            return {
                ...state,
                currentUser: action.payload,
                error: null
            }
        case GOOGLE_SIGN_IN_FAILURE:
        case EMAIL_SIGN_IN_FAILURE:
            return {
                ...state,
                error: action.payload
            }
        default:
            return state
    }
}

export default userReducer;