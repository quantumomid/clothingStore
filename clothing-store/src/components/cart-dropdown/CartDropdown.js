import React from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectCartItems } from "../../redux/cart/cartSelectors";
import { withRouter } from "react-router";
import CartItem from "../cart-item/CartItem";
import CustomButton from "../custom-button/CustomButton";
import "./CartDropdown.scss";
import { toggleCartHidden } from "../../redux/cart/cartActions";

const CartDropdown = ({ cartItems, history, dispatch }) => (
    <div className="cart-dropdown">
        <div className="cart-items" >
        {   cartItems.length ? (
            cartItems.map(cartItem => <CartItem key={cartItem.id} item={cartItem} />))
            : <span className="empty-message">Your cart is empty</span>
        }
        </div>
        <CustomButton onClick={() => {
            history.push('/checkout')
            dispatch(toggleCartHidden())   
        }}>GO TO CHECKOUT</CustomButton>
    </div>
);

const mapStateToProps = createStructuredSelector({
    cartItems: selectCartItems
});

// Note: the order of wrapping is important i.e. to pass the "connected" component
// into withRouter so that the correct end state of CartDropdown has access to the 
// various router properties we desire
export default withRouter(connect(mapStateToProps)(CartDropdown));

// Also note when the second parameter to connect() is empty then connect automatically
// providers dispatch() in the props for you to use! 