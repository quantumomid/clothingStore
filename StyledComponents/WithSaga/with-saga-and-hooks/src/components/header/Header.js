import { connect } from "react-redux";
// below is a special syntax for importing SVG in React :o
import { ReactComponent as Logo } from "../../assets/crown.svg";
import CartIcon from "../cart-icon/CartIcon";
import CartDropdown from "../cart-dropdown/CartDropdown";
import { selectCartHidden } from "../../redux/cart/cartSelectors";
import { selectCurrentUser } from "../../redux/user/userSelectors";
import { createStructuredSelector } from "reselect";

import { HeaderContainer, LogoContainer, OptionLink, OptionsContainer } from "./headerStyles";
import { signOutStart } from "../../redux/user/userActions";

function Header({ currentUser, hidden, signOutStart }){

    return(
        <HeaderContainer>

            <LogoContainer to="/">
                <Logo className="logo" />
            </LogoContainer>

            <OptionsContainer>
                <OptionLink to='/shop'>
                    SHOP
                </OptionLink>
                <OptionLink to='/shop'>
                    CONTACT
                </OptionLink>
                { currentUser ? (
                    <OptionLink as="div" onClick={signOutStart}>SIGN-OUT</OptionLink>
                ) : (
                    <OptionLink to='/signin'>
                    SIGN-IN
                </OptionLink>
                )}
                <CartIcon />
            </ OptionsContainer>
            {
                hidden ? null :
                <CartDropdown />
            }

        </HeaderContainer>
    )
}

const mapStateToProps = createStructuredSelector({
    currentUser: selectCurrentUser,
    hidden: selectCartHidden
});

const mapDispatchToProps = dispatch => ({
    signOutStart: () => dispatch(signOutStart())
})

export default connect(mapStateToProps, mapDispatchToProps)(Header);