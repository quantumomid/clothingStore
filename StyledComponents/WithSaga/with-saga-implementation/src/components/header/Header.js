import { connect } from "react-redux";
import { auth } from "../../firebase/firebase.utils";
// below is a special syntax for importing SVG in React :o
import { ReactComponent as Logo } from "../../assets/crown.svg";
import CartIcon from "../cart-icon/CartIcon";
import CartDropdown from "../cart-dropdown/CartDropdown";
import { selectCartHidden } from "../../redux/cart/cartSelectors";
import { selectCurrentUser } from "../../redux/user/userSelectors";
import { createStructuredSelector } from "reselect";

import { HeaderContainer, LogoContainer, OptionLink, OptionsContainer } from "./headerStyles";

function Header({ currentUser, hidden }){

    return(
        <HeaderContainer>

            <LogoContainer to="/">
                <Logo className="logo" />
            </LogoContainer>

            <OptionsContainer>
                <OptionLink to='/shop'>
                    SHOP
                </OptionLink>
                <OptionLink to='/shop'>
                    CONTACT
                </OptionLink>
                { currentUser ? (
                    <OptionLink as="div" onClick={() => auth.signOut()}>SIGN-OUT</OptionLink>
                ) : (
                    <OptionLink to='/signin'>
                    SIGN-IN
                </OptionLink>
                )}
                <CartIcon />
            </ OptionsContainer>
            {
                hidden ? null :
                <CartDropdown />
            }

        </HeaderContainer>
    )
}

const mapStateToProps = createStructuredSelector({
    currentUser: selectCurrentUser,
    hidden: selectCartHidden
});

export default connect(mapStateToProps)(Header);