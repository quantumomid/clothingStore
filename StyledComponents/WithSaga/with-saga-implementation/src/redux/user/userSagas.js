import { takeLatest, put, all, call } from "redux-saga/effects";
import { auth, createUserProfileDocument, googleProvider } from "../../firebase/firebase.utils";
import { emailSignInFailure, emailSignInSuccess, googleSignInFailure, googleSignInSuccess } from "./userActions";
import { EMAIL_SIGN_IN_START, GOOGLE_SIGN_IN_START } from "./userActionTypes";

export function* signInWithGoogle(){
    try {
        const { user } = yield auth.signInWithPopup(googleProvider);
        // console.log(user);
        const userRef = yield call(createUserProfileDocument, user);
        const userSnapshot = yield userRef.get();
        yield put(googleSignInSuccess({ id: userSnapshot.id, ...userSnapshot.data() }));
    } catch (error) {
        // console.log(error);
        yield put(googleSignInFailure(error));
        
    }
}

export function* onGoogleSignInStart(){
    yield takeLatest(GOOGLE_SIGN_IN_START, signInWithGoogle);
}

export function* signInWithEmail({ payload: { email, password } }){
    try {
        const { user } = yield auth.signInWithEmailAndPassword(email, password);
        const userRef = yield call(createUserProfileDocument, user);
        const userSnapshot = yield userRef.get();
        yield put(emailSignInSuccess({ id: userSnapshot.id, ...userSnapshot.data() }))
    } catch (error) {
        yield put(emailSignInFailure(error));
    }
}

export function* onEmailSignInStart(){
    yield takeLatest(EMAIL_SIGN_IN_START, signInWithEmail)
}

// Basically instantiating all the other sagas
export function* userSagas(){
    yield all([call(onGoogleSignInStart), call(onEmailSignInStart)])
}