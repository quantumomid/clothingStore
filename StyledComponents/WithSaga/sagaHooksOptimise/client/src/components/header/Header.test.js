import React from 'react';
import { BrowserRouter } from "react-router-dom";
import { mount } from 'enzyme';
import { Provider } from "react-redux";
import Header from './Header';
import { mockStore } from '../../test/testUtils';
import CartDropdown from '../cart-dropdown/CartDropdown';
import * as redux from 'react-redux';
import { signOutStart } from "../../redux/user/userActions";

const store = mockStore({});

describe('Header component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(
        <Provider store={store}>
            <BrowserRouter>
                <Header />
            </BrowserRouter>
        </Provider>);
  });

  it('Should render Header component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render sign in link when no current user', () => {
    expect(
      wrapper
        .find('OptionLink')
        .at(2)
        .text()
    ).toBe('SIGN-IN');
  });

})

describe('Header component with user logged in', () => {
    let wrapper1;
    let spyOnUseDispatch;
    let mockDispatch;
    const store1 = mockStore({
        user: {
            currentUser: {
                uid: '123'
              }
        }
    });

    beforeEach(() => {
    spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
    // Mock dispatch function returned from useDispatch
    mockDispatch = jest.fn();
    spyOnUseDispatch.mockReturnValue(mockDispatch);
      wrapper1 = mount(
          <Provider store={store1}>
              <BrowserRouter>
                  <Header />
              </BrowserRouter>
          </Provider>);
    });
  
    it('Should render sign in link when no current user', () => {
      expect(
        wrapper1
          .find('OptionLink')
          .at(2)
          .text()
      ).toBe('SIGN-OUT');
    });

    it('Should call signOutStart method when link is clicked', () => {
        wrapper1
          .find('OptionLink')
          .at(2)
          .simulate('click');
          expect(mockDispatch).toHaveBeenCalled();
          expect(mockDispatch).toHaveBeenCalledWith(signOutStart());
      });
  })

  describe('If hidden is true then CartDropdown should be showing', () => {
    const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }]
    const storeHiddenFalse = mockStore({
        cart: {
            hidden: false,
            cartItems: mockCartItems
        }
    });

    let wrapperHiddenFalse;
    
    beforeEach(() => {
      wrapperHiddenFalse = mount(
          <Provider store={storeHiddenFalse}>
              <BrowserRouter>
                  <Header />
              </BrowserRouter>
          </Provider>);
    });

    it('should not render CartDropdown', () => {
      expect(wrapperHiddenFalse.exists(CartDropdown)).toBe(true);
    });
  });