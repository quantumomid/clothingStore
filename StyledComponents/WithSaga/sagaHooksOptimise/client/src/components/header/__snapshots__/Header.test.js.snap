// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Should render Header component 1`] = `ReactWrapper {}`;
